shacl12-rules/index.html

```diff
@@ -700,163 +655,292 @@ <h3>RDF Rules Syntax</h3>
 
-eval(F(expr1, expr2), row) = F(eval(expr1, row), eval(expr2, row))
-eval(FF(expr1, expr2) , row) = ... things that are not functions like IF
+eval(F(expr1, expr2 ...), row) = F(eval(expr1, row), eval(expr2, row), ...)
+ ...
+eval(FF(expr1, expr2) , row) = ... things that are not functions like `IF`
 </pre>
       </section>
 
       <section id="eval-rule">
-        <h4>Evaluation of a rule</h4>
+        <h4>Evaluation of a Rule</h4>
+
+        <div class=ednote>
+          <p>Sketch</p>
+          <p>Turn into definitions of the step components</p>
+        </div>
+
+
 <pre class="algorithm">
 let R be a well-formed rule.
-let rule R = (head, body) where
+let rule R = (H, B) where
              H is the sequence of triple templates in the head
              B is the sequence of triple patterns, condition expressions,
                 and assignments in the body
 
 let R : map variable to RDF term as a set of pairs (variable, RDF term)
 
-# "solution"
-let T :list of B
+## Solution sequence of one solution that does not map any variables.
+## (This is the join identity)
+
+let SEQ: Solution sequence = { <var>μ</var><sub>0</sub> }
+let G = <a>evaluation graph</a>
+
+# Start::
 
-Initial T = empty list
+Replace each blank node in B with a fresh 
+variable that does not occur in the rule body.
 
 # Evaluate rule body
-for each rule element rElt:
-    if rElt is a triple pattern:
-        T1 = empty list
-    for each row in T:
-      T1 = empty list
-      for each matches m set of (var,term)
-          row1 = row union (var, term)
-          add row1 to T1
-          endfor
-    T = T1
-    endif
+for each rule element rElt in B
+    let S1 = { }
+
+    if rElt is a triple pattern TP:
+        X = graphMatch(G, TP)
+        SEQ1 = { <var>μ</var><sub>0</sub> }
+        for each <var>μ</var><sub>1</sub> in X:
+            for each <var>μ</var><sub>2</sub> in T:
+              if compatible(<var>μ</var><sub>1</sub>, <var>μ</var><sub>2</sub>)
+                <var>μ</var><sub>3</sub> = merge(<var>μ</var>1, <var>μ</var><sub>2</sub>)
+                add <var>μ</var><sub>3</sub> to SEQ1
+            endfor
+        endif
 
     if rElt is a condition expression F:
-        T1 = empty list
-        for each row in T:
-            if ( eval(F,B) )
-                add R to T1
+        SEQ1 = {}
+        for each solution <var>μ</var> in SEQ:
+            ## EBV = Effective boolean value.
+            ## TODO: Eval errors.
+            if ( EBV(eval(<var>μ</var>, F)) is true )
+                add <var>μ</var> to SEQ1
                 endif
-        endfor
-    T = T1
-    endif
+            endfor
+        endif
 
     if rElt is an assignment(V, expr)
-        T1 = empty list
-        for each row in T:
-            let x = eval(expr, row)
-            let R1 = R union (V, x)
-            add R1 to T1
+        SEQ1 = {}
+        for each solution S in SEQ:
+            let x = eval(expr, <var>μ</var>)
```

Comment:
`μ` is undefined here, shouldnt it be S?
```suggestion
            let x = eval(expr, S)
```